cpp/src/interop/to_arrow_device.cu

```diff
@@ -0,0 +1,590 @@
+  return cudf::type_dispatcher(
+    dict_keys.type(),
+    detail::dispatch_to_arrow_type{},
+    dict_keys,
+    dict_keys.type().id(),
+    metadata.children_meta.empty() ? column_metadata{"keys"} : metadata.children_meta[0],
+    out->dictionary);
+}
+
+template <typename T>
+static void device_buffer_finalize(ArrowBufferAllocator* allocator, uint8_t*, int64_t)
+{
+  auto* unique_buffer = reinterpret_cast<std::unique_ptr<T>*>(allocator->private_data);
+  delete unique_buffer;
+}
+
+template <typename T>
+void set_buffer(std::unique_ptr<T>& device_buf, int64_t i, struct ArrowArray* out)
+{
+  ArrowBuffer* buf = ArrowArrayBuffer(out, i);
+  buf->data        = reinterpret_cast<uint8_t*>(device_buf->data());
+  buf->size_bytes = device_buf->size();
+  // we make a new unique_ptr and move to it in case there was a custom deleter
+  ArrowBufferSetAllocator(buf,
+                          ArrowBufferDeallocator(&device_buffer_finalize<T>,
+                                                 new std::unique_ptr<T>(std::move(device_buf))));
+}
+
+struct dispatch_to_arrow_device {
+  template <typename T, CUDF_ENABLE_IF(not is_rep_layout_compatible<T>())>
+  int operator()(cudf::column&,
+                 cudf::type_id,
+                 rmm::cuda_stream_view,
+                 rmm::mr::device_memory_resource*,
+                 struct ArrowArray*)
+  {
+    CUDF_FAIL("Unsupportd type for to_arrow_device");
+  }
+
+  template <typename T, CUDF_ENABLE_IF(is_rep_layout_compatible<T>())>
+  int operator()(cudf::column& column,
+                 cudf::type_id id,
+                 rmm::cuda_stream_view stream,
+                 rmm::mr::device_memory_resource* mr,
+                 struct ArrowArray* out)
+  {
+    nanoarrow::UniqueArray tmp;
+
+    ArrowType storage_type;
+    switch (id) {
+      case cudf::type_id::TIMESTAMP_SECONDS:
+      case cudf::type_id::TIMESTAMP_MILLISECONDS:
+      case cudf::type_id::TIMESTAMP_MICROSECONDS:
+      case cudf::type_id::TIMESTAMP_NANOSECONDS: storage_type = NANOARROW_TYPE_TIMESTAMP; break;
+      case cudf::type_id::DURATION_SECONDS:
+      case cudf::type_id::DURATION_MILLISECONDS:
+      case cudf::type_id::DURATION_MICROSECONDS:
+      case cudf::type_id::DURATION_NANOSECONDS: storage_type = NANOARROW_TYPE_DURATION; break;
+      default: storage_type = id_to_arrow_type(id);
+    }
+
+    NANOARROW_RETURN_NOT_OK(ArrowArrayInitFromType(tmp.get(), storage_type));
+    tmp->length     = column.size();
+    tmp->null_count = column.null_count();
+
+    auto contents = column.release();
+    if (contents.null_mask) { set_buffer(contents.null_mask, 0, tmp.get()); }
+
+    set_buffer(contents.data, 1, tmp.get());
+
+    NANOARROW_RETURN_NOT_OK(
+      ArrowArrayFinishBuilding(tmp.get(), NANOARROW_VALIDATION_LEVEL_MINIMAL, nullptr));
+    ArrowArrayMove(tmp.get(), out);
+    return NANOARROW_OK;
+  }
+};
+
+template <typename DeviceType>
+int unsupported_decimals_to_arrow(cudf::column& input,
```

Comment:
I was trying to avoid changing or touching the existing `to_arrow` / `from_arrow` stuff in order to ensure no potential changes in behavior for those functions and that unification and refactoring could be done as a second pass afterwards. But I can look into doing this conversion.

Reply:
Yup I agree. Let's do it in a follow-up.